### PR TITLE
[LowerToHW] Wires used for output ports should get their symbols.

### LINKED ISSUE #5011: [FIRRTL] LowerToHW replaces uses of output ports and doesn't preserve dontTouch (symbol)
More unexpected behavior with output ports, dontTouch, and reading them within a module.

Similar to #5008 .

Consider this FIRRTL input:

```firrtl
circuit Producer : %[[
 { "class": "firrtl.transforms.DontTouchAnnotation",
   "target": "~Producer|Producer>out" },
 { "class": "firrtl.transforms.DontTouchAnnotation",
   "target": "~Producer|Producer>n" }
]]
  module Producer:
    input clock: Clock
    output out: UInt<5>

    ; Simple constant output.
    node n = UInt<5>(1)
    out <= n

    assert(clock, neq(out, UInt<5>(3)), UInt<1>(1), "out was changed")
```

Which w/`-ir-fir` we get (good so far):

```mlir
module {
  firrtl.circuit "Producer" {
    firrtl.module @Producer(in %clock: !firrtl.clock, out %out: !firrtl.uint<5> [{class = "firrtl.transforms.DontTouchAnnotation"}]) attributes {convention = #firrtl<convention scalarized>} {
      %c1_ui1 = firrtl.constant 1 : !firrtl.uint<1>
      %c3_ui5 = firrtl.constant 3 : !firrtl.uint<5>
      %c1_ui5 = firrtl.constant 1 : !firrtl.uint<5>
      %n = firrtl.node %c1_ui5 {annotations = [{class = "firrtl.transforms.DontTouchAnnotation"}]} : !firrtl.uint<5>
      firrtl.strictconnect %out, %n : !firrtl.uint<5>
      %0 = firrtl.neq %out, %c3_ui5 : (!firrtl.uint<5>, !firrtl.uint<5>) -> !firrtl.uint<1>
      firrtl.assert %clock, %0, %c1_ui1, "out was changed" : !firrtl.clock, !firrtl.uint<1>, !firrtl.uint<1>  {eventControl = 0 : i32, isConcurrent = false}
    }
  }
}
```

Piping the above through `circt-opt --lower-firrtl-to-hw` yields:

```mlir
module {
  hw.module @Producer(%clock: i1) -> (out: i5 {hw.exportPort = #hw<innerSym@__Producer__out>}) {
    %c1_i5 = hw.constant 1 : i5
    %c3_i5 = hw.constant 3 : i5
    %true = hw.constant true
    %.out.output = hw.wire %n  : i5
    %n = hw.wire %c1_i5 sym @__Producer__n  : i5
    %0 = comb.icmp bin ne %.out.output, %c3_i5 : i5
    sv.always posedge %clock {
      sv.if %true {
        sv.assert %0, immediate message "out was changed"
      }
    }
    hw.output %.out.output : i5
  }
}
```

Where the use of the output port is replaced with a wire (`%.out.output`) which lacks the "dontTouch"-y symbol.

As a result, running this input through the pipeline currently produces:

```systemverilog
// Generated by CIRCT 1.37.0g20230411_c9c3591
module Producer(
  input        clock,
  output [4:0] out
);

  wire [4:0] n = 5'h1;
  always @(posedge clock)
    assert(n != 5'h3) else $error("out was changed");
  assign out = n;
endmodule
```

### PR DESCRIPTION
**Pull request by dtzSiFive**
Fixes #5011.

### CODE REVIEW

**Comment by dtzSiFive** [comment]
Draft as builds on #5724 .

**Comment by dtzSiFive** [comment]
This means the symbol will track the wire, not the port, which may have implications re:metadata.  However, it also ensures "dontTouch" and use-cases interested in tracking the port functionally are better served by following the wire (for example, use in force targets).

**Review by fabianschuiki — APPROVED** [review]
LGTM!

Could this be some precedence for a future where ports are just a mapping from call-site to callee SSA values, without them themselves being any kind of wire or forceable thing. And all the interaction / wire-ness would be added by dedicated ops before or after the port boundary :smirk: ?

**Comment by dtzSiFive** [comment]
> Could this be some precedence for a future where ports are just a mapping from call-site to callee SSA values, without them themselves being any kind of wire or forceable thing. And all the interaction / wire-ness would be added by dedicated ops before or after the port boundary smirk ?

Sounds great to me! :smirk_cat:

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp [hunk 1]
```diff
@@ -1371,18 +1371,25 @@ FIRRTLModuleLowering::lowerModuleBody(FModuleOp oldModule,
...
     // Don't output zero bit results or inouts.
     auto resultHWType = loweringState.lowerType(port.type, port.loc);
     if (!resultHWType.isInteger(0)) {
-      auto output = castFromFIRRTLType(newArg, resultHWType, outputBuilder);
+      auto output =
+          castFromFIRRTLType(newArg.getResult(), resultHWType, outputBuilder);
+      auto idx = newModule.getNumInputs() + outputs.size();
       outputs.push_back(output);
+
+      // If output port has symbol, move it to this wire.
```
Comment (darthscsi) [inline lib/Conversion/FIRRTLToHW/LowerToHW.cpp:1388]: What about "if output port had a don't touch, add a symbol to this wire also"?

Comment (dtzSiFive) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp:1388]: I don't remember quite but as I recall whether the original has dontTouch is no longer available in current approach, FWIW.

More to the point, that'd be a hack for the dontTouch case specifically and doesn't address the fundamental incompatible interpretations of symbols as structural (name this port) vs functional (name this point in the dataflow), which requires moving the symbol not spawning new ones.

**Pull request merged**